File: javascript/exponent.js
function exp(base, n) {
	if (n === 0) { return 1 }
	return base * exp(base, n - 1);
}

function exp2(base, n) {
  if (n === 0) { return 1 }
	var num;
  if (n % 2 === 0) {
    num = exp2(base, n/2);
    return num * num;
  } else {
    num = exp2(base, (n - 1) / 2);
    return base * num * num;
  }
}

// console.log(exp(3, 8));
// console.log(exp2(3, 8));

File: ruby/exponent.rb
def exponent1(base, n)
  return 1 if n == 0
  return base * exponent1(base, n - 1)
end

def exponent2(base, n)
  return 1 if n.zero?
  if n.even?
    num = exponent2(base, n/2)
    return num * num
  else
    num = exponent2(base, (n - 1) / 2)
    return base * num * num
  end
end
